Stowaway Changelings (Latejoin Changelings)

iwishforducks opened the pull request:
## About The Pull Request

Adds a latejoin form of Changeling flavored as a stowaway.

## Why It's Good For The Game

We don't have a latejoin form of Changelings. It ups the paranoia that anyone can be a Changeling, latejoin or not. They're also one of the few non-progression antagonists which would work perfectly as a late join.

## Changelog

:cl:
add: Stowaway Changelings will now appear as a late join form of Changeling.
/:cl:

iwishforducks committed:
Stowaway Changelings

Adds a latejoin form of Changeling flavored as a stowaway.

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -216,3 +216,41 @@
 
 	return TRUE
+
+//////////////////////////////////////////////
+//                                          //
+//           LATEJOIN CHANGELING            //
+//                                          //
+//////////////////////////////////////////////
+
```
```suggestion
/// Ruleset for latejoin changelings
```

These ascii art style comments are banned because they break autodoc and are ugly

iwishforducks replied:
All of the dynamic rulesets have these ascii art styles. Do you want me to go through and fix them all after this one?

iwishforducks replied:
![image](https://user-images.githubusercontent.com/65363339/231333680-aaae30ac-faf4-4013-a122-680b1c85f6e7.png)

Mothblocks commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -216,3 +216,41 @@
+	var/mob/picked_mob = pick(candidates)
+	assigned += picked_mob.mind
+	picked_mob.mind.special_role = antag_flag
+	var/datum/antagonist/heretic/new_changeling = picked_mob.mind.add_antag_datum(antag_datum)
```
`heretic`?

Also why do you need to save this as a variable?

iwishforducks replied:
No reason to save it as a variable you're correct. Just some lame copy+paste code I didn't look at too carefully.

iwishforducks commented:
> > We don't have a latejoin form of Changelings.
> 
> Would like some more description on this because this isn't an oversight but a deliberate decision to limit the amount of changelings. Not against it anymore but would like this to not be treated as innately beneficial

You're right, I had more reasoning to go with it but I'm not sure why I didn't put it there. Just updated it.

Mothblocks commented on code/modules/client/preferences/migrations/tgui_prefs_migration.dm:
```diff
@@ -39,6 +39,7 @@
 	migrate_antagonist(ROLE_REV_HEAD, list(ROLE_PROVOCATEUR))
 	migrate_antagonist(ROLE_TRAITOR, list(ROLE_SYNDICATE_INFILTRATOR, ROLE_SLEEPER_AGENT))
 	migrate_antagonist(ROLE_WIZARD, list(ROLE_WIZARD_MIDROUND))
+	migrate_antagonist(ROLE_CHANGELING, list(ROLE_STOWAWAY_CHANGELING))
```
You don't need to do this because this was only for preferences that existed before the new prefs menu, when traitor meant all forms of traitor, etc. Because there was no latejoin changeling, nobody who picked changeling intended to play it. This should be treated as immutable and won't update existing players anyway.

Fikou commented on tgui/packages/tgui/interfaces/PreferencesMenu/antagonists/antagonists/stowawaychangeling.ts:
```diff
@@ -0,0 +1,18 @@
+  description: [
+    multiline`
+      A Changeling that found its way onto the shuttle
+      unbeknowst to the crewmembers on board.
```
```suggestion
      unbeknownst to the crewmembers on board.
```

Mothblocks commented:
> You're right, I had more reasoning to go with it but I'm not sure why I didn't put it there. Just updated it.

:+1: Looks fine to me now

iwishforducks committed:
Grammar check

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

iwishforducks committed:
Erroneous variable, pref migration removal

Removes erroneous variable I was storing + calling it a heretic (?)
Removes preference migration code as it's not needed

iwishforducks committed:
Merge branch 'latejoin-lings' of https://github.com/iwishforducks/tgstation into latejoin-lings

iwishforducks committed:
ascii art begone

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

timothymtorres commented:
We already have the midround changeling meteor event from #73018.  Why do we need both?

Mothblocks commented:
Hmm yeah that is a decent point, what do you think @iwishforducks

carlarctg commented:
should they start with bonus shop points? latejoining puts you at quite the disadvantage vs prepared crew, and sleeper agents get rep scaling to the round time

iwishforducks commented:
> We already have the midround changeling meteor event from #73018. Why do we need both?

midround changeling is very unique and plays much differently from a latejoin changeling. i don't want to take away its uniqueness at all. this latejoin changeling is meant to fill in the gap of dynamic not having enough good latejoin antagonists to choose from. i think both are perfectly capable of being in circulation: much like how we have nuclear assault midrounds and lone operative midrounds

iwishforducks commented:
also right now if someone comes from the arrival shuttle then you have no reason to fear them being a changeling. midround changeling will rarely ever fill this gap.

Mothblocks commented:
> I think we should maybe make one or the other cancel each other out, like disable the random event when this rolls, and disable this antagonist when the random event rolls (you can do both of these by adjusting the weight)

do you think you can do this reuqest then

MrStonedOne commented:
I don't like making events and dynamic mess with eachother

It adds to the round to round chaos that these are two seperate things that can stack.

or to put it another way, dynamic rolls is an OOC injection of chaos into the round, events are an IC injection of chaos into the round, and I like the idea of them working entirely in parallel.

Mothblocks commented:
dynamic impacting events is already very precedented with antag events being basically disabled right before/after dynamic rolls (+/- 15 minutes)

the intent of this non-feature is less changelings, im ok with some more changelings, but if we've already done that then i'm already satisfied and without that change i don't think i want this

iwishforducks commented:
> > I think we should maybe make one or the other cancel each other out, like disable the random event when this rolls, and disable this antagonist when the random event rolls (you can do both of these by adjusting the weight)
> 
> do you think you can do this reuqest then

no. not for the reasons MSO outlined- but I do agree with him- but because our midround changelings are much more different than the proposed latejoin changeling. i explicitly do not want to step on the toes of the midround changeling because i think it's a fantastic antagonist. if we're concerned about the amount of changelings on the field then we should be adjusting the weights of the latejoin and midround respectively.

Mothblocks commented:
i probably need to do my work on splitting weights again then because right now the system is sstill not comprehensible

jlsnow301 committed:
Merge branch 'master' into latejoin-lings

ModDownloading commented:
I agree that I think the changeling meteor event is probably a better choice for this. We don't really need two options for more lings to show up.

nukularpower commented:
I would disagree that we "dont need more options" for changelings - I think this would be a fine addition and play quite differently than the meteorling.   Also, as mentioned by the submitter, this would address the current issue with Dynamic much better.

+1 from me!

Mothblocks commented:
you need to fix the screenshot tests

iwishforducks committed:
screenshots for screenshot tests

iwishforducks commented:
sorry took me a while to get back to this but should be fixed now @Mothblocks

Mothblocks commented:
Undraft if ready

JohnFulpWillard commented on tgui/packages/tgui/interfaces/PreferencesMenu/antagonists/antagonists/stowawaychangeling.ts:
```diff
@@ -0,0 +1,18 @@
+  name: 'Stowaway Changeling',
+  description: [
+    multiline`
+      A Changeling that found its way onto the shuttle
+      unbeknownst to the crewmembers on board.
```
```suggestion
      A Changeling surprise replacement one of the
      crewmembers during their transit to the station.
```
my wording is poor but i think implying they killed the person instead would be more flavorful, up to you though

iwishforducks replied:
Methinks the original wording can still be interpreted as such. That was my original intention anyhow, to allow it to either be interpreted as a head ling or a full fledged ling
